src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java

```diff
@@ -0,0 +1,336 @@
+        if (!oReplicateWithLatestChanges.isPresent()) return null;
+        
+        replicate = oReplicateWithLatestChanges.get();
+
+        boolean didReplicateReveal = replicate.getLastRelevantStatus()
+                .equals(ReplicateStatus.REVEALED);
+
+        boolean wasReplicateRequestedToUpload = replicate.getLastRelevantStatus()
+                .equals(ReplicateStatus.RESULT_UPLOAD_REQUESTED);
+
+        if (didReplicateReveal) {               
+            return RecoveryAction.WAIT;
+        }
+
+        if (wasReplicateRequestedToUpload) {
+            return RecoveryAction.UPLOAD_RESULT;
+        }
+
+        return null;
+    }
+
+    /**
+     * RESULT_UPLOAD_REQUESTED          => RecoveryAction.UPLOAD_RESULT
+     * RESULT_UPLOADING + !done yet     => RecoveryAction.UPLOAD_RESULT
+     * RESULT_UPLOADING + done          => RecoveryAction.WAIT
+     *                                     update to ReplicateStatus.RESULT_UPLOADED
+     * RESULT_UPLOADED                  => RecoveryAction.WAIT 
+     */
+
+    private RecoveryAction recoverReplicateInResultUploadPhase(Task task, Replicate replicate) {
+        String chainTaskId = task.getChainTaskId();
+        String walletAddress = replicate.getWalletAddress();
+
+        boolean wasReplicateRequestedToUpload = replicate.getLastRelevantStatus().equals(ReplicateStatus.RESULT_UPLOAD_REQUESTED);
+        boolean didReplicateStartUploading = replicate.getLastRelevantStatus().equals(ReplicateStatus.RESULT_UPLOADING);
+        boolean didReplicateUploadWithoutNotifying = replicatesService.hasResultBeenUploaded(task.getChainTaskId());
+        boolean hasReplicateAlreadyUploaded = replicate.getLastRelevantStatus().equals(ReplicateStatus.RESULT_UPLOADED);
+
+        if (wasReplicateRequestedToUpload) {
+            return RecoveryAction.UPLOAD_RESULT;
+        }
+
+        if (didReplicateStartUploading && !didReplicateUploadWithoutNotifying) {
+            return RecoveryAction.UPLOAD_RESULT;
+        }
+
+        if (didReplicateStartUploading && didReplicateUploadWithoutNotifying) {
+            replicatesService.updateReplicateStatus(chainTaskId, walletAddress,
+                    ReplicateStatus.RESULT_UPLOADED, ReplicateStatusModifier.POOL_MANAGER);
+
+            taskExecutorEngine.updateTask(chainTaskId);
+            return RecoveryAction.WAIT;
+        }
+
+        if (hasReplicateAlreadyUploaded) {
+            return RecoveryAction.WAIT;
+        }
+
+        return null;
+    }
+
+    /**
+     * REVEALED          => RecoveryAction.COMPLETE
+     * !REVEALED         => null
+     */
+
+    private RecoveryAction recoverReplicateIfRevealed(Replicate replicate) {
+        // refresh task
+        Optional<Task> oTask = taskService.getTaskByChainTaskId(replicate.getChainTaskId());
+        if (!oTask.isPresent()) return null;
+
+        return replicate.containsRevealedStatus()
```

Comment:
Nice one but to difficult to read, can we simplify a bit? :)

Reply:
of course, also done!